### ISSUE TITLE
NUnit project file settings are ignored

### ISSUE THREAD

**Issue by cyanite** [nunit/nunit-console#730]
The way TestPackages deal with settings means that settings always propagate down to sub packages, but never up. At launch, all console line settings are set on the top level package, and are propagated down to the next package, which could be an .nunit project.

This subpackage later (at expand time) loads its settings, such as process model and domain model, from the project file. These settings are never propagated back up to the top level TestPackage, however, so they never take proper effect.

This was already a problem in 3.10, but it still managed to work because NUnit would (incorrectly, I think) create an AggregateTestRunner for the single .nunit project, which would then in turn instantiate the correct runner (say a multi process version).

Since the aggregate test runner has been removed, this is no longer the case and NUnit will always use a single process runner even when the .nunit project specifies otherwise. In my case, this causes all my tests to fail.

So, this is a regression since 3.10, but in my opinion the settings handling was already broken earlier.

The simplest fix is this: Whenever GetSetting is called and we have exactly one sub package, forward the call to the sub package instead of handling it ourselves. I've created a pull request for this fix.

If you'd prefer a different fix that's fine, but it will leave us in the odd situation of taking action based on only console settings in several places.

**Comment by cyanite** [comment]
It's unfortunate that I didn't debug this before the 3.11 release.

**Comment by cyanite** [comment]
Other things are broken as well by this bug, I think, such as selecting the correct default configuration. This is also read on the top level package, but will always have the default value (or from command line options) there.

**Comment by cyanite** [comment]
An alternative approach is to only do this at the top level, although this requires some further code change since a TestPackage currently doesn't know if it's the top level package.

**Comment by CharliePoole** [comment]
UPDATE: OK... on the first reading, I missed "when we have exactly one subpackage." So I see you are already concerned about the problem of conflicting settings. However, I'll leave my more general comments because, as you will see, "exactly one subpackage" is not a sufficient qualification.

The engine passes settings downward if they are added with AddSetting. If you use Settings.Add, they only apply to the specific package.

So why (and when) do we want to pass settings downward? Answer: when we are overriding settings from the command-line. The command-line is the only place where we create settings and pass them downward, because the meaning of the options like --process is that they override individual settings, which would otherwise be used, on specific assemblies.

So where do the "natural" settings for each assembly come from? That is, the settings we are overriding?

1. From the assembly itself. The engine analyzes it and sets up things like target runtime, whether x86 is required, etc. If we propagate those things upward we could run into conflicts __and__ we make the top level appear as if the user specified an override.

2. From projects, which may supply their own settings for all contained assemblies or for certain assemblies. The `.nunit` project does that. Upward propogation runs the same risks as point 1.

3. In some environments, the user may be permitted to change the settings of individual packages. For example, a GUI can allow a context menu for making such settings changes.

As you can see, the tree of `TestPackages` is intended to allow arbitrary settings changes, both upward and downward. But if you want that detailed information, you have to dig through the tree. 

Aside from all that, the idea of making a change to the semantics of `GetSetting()`, right in the api assembly, strikes me as breaking in a rather dangerous way. That is, it __doesn't__ change the syntax, so everything will compile well, but the method will do something different.

`AggregateTestRunner` has not been removed but it has been moved. It's now in the `nunit.engine.core` assembly, which is referenced by `nunit.engine`. It's possible that some changes have been made as to when `AggregateTestRunner` is used... I didn't check that part of the latest code. The original intent of using `AggregateTestRunner` for projects was that we didn't know (at the time of assigning the runner) what the package contained. I'm not sure how that works now but maybe @ChrisMaddock can reply as I don't have any more time at the instant.

As far as I can see, you haven't specified exactly what the regression is... i.e. what problem you want to solve. Can you do that with an example?

BTW, I faced problems of needing more or different things from test packages than what the API includes in working on TestCentric. I resolved my issues by creating extension methods. The ones I use are available here: https://github.com/TestCentric/testcentric-gui/blob/master/src/TestEngine/testcentric.engine.core/Helpers/TestPackageExtensions.cs

**Comment by cyanite** [comment]
> As far as I can see, you haven't specified exactly what the regression is... i.e. what problem you want to solve. Can you do that with an example?

.nunit settings are completely ignored. So processModel, domainModel and default setting are never used, since they are set on the subpackage.

The main package now says: "let me see what process model to use".. "not found.. .so default". And then "which runner to use with 'default' and just one sub package?" "single process runner!"

Before instead of single process is here picked aggregate which saved the situation somewhat (although the default setting was still ignored, I think.)

In short: you can't specify settings in an nunit file anymore.

**Comment by cyanite** [comment]
> The command-line is the only place where we create settings and pass them downward,

They are passed downward on all levels, though, which means that nunit file settings are passed down to their assemblies.

**Comment by CharliePoole** [comment]
@cyanite 

Re "top-level"

You __could__ detect top-level by the fact that the name is null. That's the intention but unfortunately it's not enforced anywhere currently. That would still present the problem of re-defining an API, albeit in only some circumstances.

Re: "exactly one"

What you appear to want is "exactly one assembly" which is of course different from "exactly one subpackage". I did that with an extension method to retrieve a list of assembly packages. Of course, heuristics are involved because the packages don't identify themselves cleanly as "top level", "project" or "assembly".

**Comment by cyanite** [comment]
> What you appear to want is "exactly one assembly" which is of course different from "exactly one subpackage"

Well, what I want is that when I type `nunit3-console test.nunit`, the settings in the nunit file are not ignored. One way is to grab settings from a sub package if there is just one, at least on the top level (since the top level will always be: top test package -> nunit file (or multiple, or assemblies) -> assemblies etc.)

Now, if you specify *more* than one nunit file of course we can't propagate settings up, but in that case you get a multi runner anyway.

**Comment by CharliePoole** [comment]
@cyanite (All my comments should be checked with @ChrisMaddock, since I'm not doing a lot of work on this engine lately.)

If .NUnit settings are ignored, that's a bug in the nunit project loader. Project loaders actually create the package for the project. This used to work. If not, someone should fix that rather than changing how the TestPackage API works - at least IMO!

Something I am not understanding... (this relates to specifying the __problem__ :wink: )...

Are you wanting the API semantics to change because __you__ have code that calls it? Or are you wanting it to change because of how the console currently calls it? The solutions are different in each case, obviously.

If __you__ want a different API, I'd have to say (again, my opinion only) we can't do that.

If you merely want the extension to work differently from how it does, we should file a bug against it. I'm not sure who will fix it, of course, but that seems (again) an obvious solution.

If we want the __console__ to work differently, we should fix the console. There are many runners of NUnit tests and it would be a bad thing to change the engine API just for the console and (possibly) break some other runner - especially my runner. :wink:

**Comment by CharliePoole** [comment]
@cyanite Re: NUnit project settings

"They are passed downward on all levels, though, which means that nunit file settings are passed down to their assemblies."

But elsewhere, you indicated that the settings in the `.nunit` file are ignored, so I'm confused. Can you clarify in what sense they are ignored? Here's what is supposed to happen...

1. If the NUnit file calls for a single separate process and you specify no command-line options, that's what you should get.

2. If the NUnit file calls for a single separate process and you override it at the command-lnine, then the override is what you should get.

IOW, the "default" setting is supposed to mean that no override occurs.

**Comment by cyanite** [comment]
> If .NUnit settings are ignored, that's a bug in the nunit project loader. Project loaders actually create the package for the project. This used to work. If not, someone should fix that rather than changing how the TestPackage API works - at least IMO!

The loader is fine. It loads settings into its own TestPackage, and that's the only thing it can do. The top level, however, never uses them, since it determines that it should just use a SingleProcessRunner because there is only one subpackage (even though that subpackage is an nunit file which says otherwise.)

> If you want a different API, I'd have to say (again, my opinion only) we can't do that.

Look, I hate to be blunt but you (the general you) broke the product in a pretty bad way, a product thousands of people rely on. It no longer reads settings in its own project format. That's a huge regression. Yet here we are discussing *my* wishes. My wish is to fix the bug, nothing more.

**Comment by CharliePoole** [comment]
OK, I'll make one more try and then back off. Maybe you can explain better or someone who actually worked on the release can take over and will know what you are talking about.

This thread has gone in several directions because I was trying to guess what your problem is. That's a fault I have and it tends to come into play when I'm presented with a solution rather that a problem statement. I'm stopping, at least for this thread.

So far, here is what I know... 
 * Something that worked before doesn't work any more. You haven't said precisely what that something is.
 * You want to fix it by introducing a new breaking change in the Test PackageAPI. I get that the change fixes your problem, but that doesn't tell me what the problem was in the first place.

BTW, I'm not part of that "general you" because I didn't work on the release. IF I had, maybe I would know what you're talking about. However, I have checked to make sure that the TestPackage API did __not__ change.

@cyanite If it's in you to explain more clearly and exactly what broke on you, that will help. Something like "When I use setting X in the `.nunit` file, the console runner overrides it and uses setting Y."

@nunit/engine-team If anyone understands better than I do what the problem is, please step in. I only chimed in out of fear that y'all might accept the proposed change, which I think is a bad idea.

**Comment by cyanite** [comment]
> Something that worked before doesn't work any more. You haven't said precisely what that something is.

It's in the title of the bug? I've also repeated it several times. Here goes: Settings inside NUnit files are now ignored. This includes at least processModel and domainUsage, and might include activeconfig as well, in which case the latter was broken in 3.10 as well.

> You want to fix it by introducing a new breaking change in the Test PackageAPI.

Well, I wrote explicitly that it was the quickest fix I could think of and that we could fix it in some other way. It says so at the end of the description. You claim it's breaking, but it doesn't break any tests, and you haven't provided an example where it breaks something. That said, I agree that it might break something, although I can't see what.

> BTW, I'm not part of that "general you"

But you're the one replying :p

>  If it's in you to explain more clearly and exactly what broke on you, that will help. Something like "When I use setting X in the .nunit file, the console runner overrides it and uses setting Y."

I can only explain it so many times. It's in the title, it's in the main description several times, and I made comments repeating it. Nunit-console3 ignores settings in its own nunit files, which is a major breaking change and clearly a bug.

**Comment by CharliePoole** [comment]
Sorry to frustrate you. I took your title plus description to mean that settings in project files are not used because of settings not propagating upward. If that's not what you meant, then I can see my comments won't make sense.

Breaking changes have to do with other people's code. I'm a user like you. If I had upgraded to the new release I'd probably be complaining too. But I'll also be unhappy if the API changes, because I know I rely on it to run my own tests. 

Let's wait for a team response.

**Comment by ChrisMaddock** [comment]
Hey @cyanite - thanks for digging into this. I've repro'd the issue you've raised. Can I just confirm that you can workaround this at the moment by adding `--process=multiple` on the command line? Appreciate that that would be a workaround, and not a solution!

---

This part of the engine is a bit of an architectural nightmare - and Charlie and I have both done an amount of work in 3.11 to try and rationalise things. There's still a bunch to do here, you can see all the related problems we want to tackle in the [3.12 milestone](https://github.com/nunit/nunit-console/milestone/10). Unfortunately, I personally at least have had to put this on hold for a little while, to tackle .NET Core support. We'd love some help, if you're willing? 

The fix you've proposed in #731, I'm a little concerned about how that will work in all situations. For example, what if you are passing in two nunit projects on the command line - one that should run `inprocess` and another with `process=multiple`? Beyond that, I'd also prefer to fix this in a way that fits in with the overall architecture (i.e. by using the appropriate ITestRunner), rather than adding in a 'workaround' for a particular situation.

My rough current thinking is that we need to re-introduce the AggregatingTestRunner for this situation. I had a quick bash at this by making the below change in `DefaultTestRunnerFactory`, which fixed the particular issue you see.

```c#
        public override ITestEngineRunner MakeTestRunner(TestPackage package)
        {
            ProcessModel processModel = GetTargetProcessModel(package);

            switch (processModel)
            {
                default:
                case ProcessModel.Default:
                    //In this case, we have multiple layers of packages (e.g. .nunit projects) which may
                    //contain their own settings that affect runner selection. Using the AggregatingTestRunner
                    //delegates this choice to a later point.
                    if (package.SubPackages.Any(p => p.HasSubPackages()))
                        return new AggregatingTestRunner(this.ServiceContext, package);
                    else if (package.SubPackages.Count > 1)
                        return new MultipleTestProcessRunner(this.ServiceContext, package);
                    else
                        return new ProcessRunner(this.ServiceContext, package);
                   [...]
```

This change breaks a number of the unit tests in RunnerSelectionTests - but I haven't yet looked into whether those are changes which genuinely break functionality, or just change behaviour. It would also be great if we could build further on the RunnerSelectionTests to include your usecase, to prevent any regressions here in future.

Is this something you'd be willing to dig into further, along these lines? I'd be more that happy to provide thoughts along the way - when time allows. 😁

**Comment by cyanite** [comment]
Hi @ChrisMaddock,

> Can I just confirm that you can workaround this at the moment by adding --process=multiple on the command line? Appreciate that that would be a workaround, and not a solution!

Yes, this works, I was already aware, because those options are set on the top level TestPackage. It's a slightly tricky work around for us since we activate nunit3-console via various other tools.

> The fix you've proposed in #731, I'm a little concerned about how that will work in all situations. For example, what if you are passing in two nunit projects on the command line - one that should run inprocess and another with process=multiple?

Yes, I'm also a bit worried about the fix, it was just the quickest I could think of. Note that my fix only activates when there's a *single* sub package, so it wouldn't kick in in that situation, and I guess you'd be left with the multi process runner.

> My rough current thinking is that we need to re-introduce the AggregatingTestRunner for this situation.

Yes, I agree. But in that case we might as well use the AggregatingTestRunner at top level all the time (as long as there are no overriding command line options, i.e. the default case).

The current code counts the number of subpackages at the top level and makes decisions if there is just one, but this is not correct behavior since that single subpackage can select any type of runner and domain, etc.

> Is this something you'd be willing to dig into further, along these lines? I'd be more that happy to provide thoughts along the way - when time allows.

Yes, I'll see what I can do. I'm most likely able to spend some time on it.

**Comment by CharliePoole** [comment]
Hmm... contrary to what I originally told @cyanite, it seems I did have something to do with this problem because of a change I contributed back in May, 2019! In fact, there's a lot of stuff I had worked on during my year of traveling, which I see was not released until now, so I'll have to be more careful before making general statements.

Reading through the discussion between me and @ChrisMaddock on PR #622, it appears that we knew there was a problem around project settings, which someone would come back and look at in a separate PR. It doesn't look as if that ever happened. We probably should have created an issue that said "Make sure this is fixed before any release happens." 

FYI, in my own engine build for the GUI, I ensure that projects are all expanded before creating a test runner and I select the runner based on the number of assemblies found rather than the number of subpackages. However, I think I may be doing the same thing... i.e. applying the default at the top level, rather than waiting till I get down the tree... so I'll have to create an issue as well.

**Comment by cyanite** [comment]
I created a new pull request which implements @ChrisMaddock 's suggestion. This works, of course, but as you can see it requires changing a lot of tests.

However, I can't help feeling that the underlying architecture is somehow wrong or at least mismatched. It feels like we should always use the AggregatingTestRunner and not bother with all this figuring out which runner to use at the top level. Why? Well,

1. Any top-level decision can only be impacted by what we can shallowly see, plus command line options. Nothing like nunit project file options will be visible.
2. AggregatingTestRunner will have to re-make those decisions at lower levels anyway, which seems like a duplication of effort: We make decisions at top-level, but if we just picked AggregatingRunner, it will make those decisions again (and more correctly, possibly) at lower levels.

At any rate, I think this bug is important enough to fix fast. You may not agree but consider this: It breaks reading settings from the main supported project format, .nunit files. I'd classify that as a major regression. Regardless, we *have* to fix this in a local version until you release a fix, since a work-around with command line options isn't very feasible for us to implement across our test infrastructure.

Let me know if you prefer anything done differently :)

**Comment by ChrisMaddock** [comment]
Thanks for digging into this. Agree on going for the simple fix first, and, unfortunately - agree we might have to go with a minor version release to fix this. My main reason for not jumping to this immediately was: 

1) nunit files are pretty low-usage, excluding the super-simple ones generated by TeamCity
2) These options have never really worked properly, and there's a workaround available. (I'm intrigued as to why adding a command line flag isn't feasible for your test infrastructure...but maybe that's for another conversation. 🙂)

It's unfortunate that our release process is so time consuming. 😩 Will take a proper look at things at the weekend.

**Comment by cyanite** [comment]
Thanks @ChrisMaddock 

> nunit files are pretty low-usage, excluding the super-simple ones generated by TeamCity

Ah ok, I didn't know that. We use them pretty extensively, but I can see that it maybe mostly makes sense for large-ish monoliths or similar. We have hundreds of assemblies, so we have a harvesting system to create the nunit files. It's also important for us to control the domain/process model, due to the way we use the dll loading system. A few of my earlier contributions to NUnit has been to fix edge cases in that behavior 🙂.

> These options have been buggy for a long time, and there's a workaround available. (I'm intrigued as to why adding a command line flag isn't feasible for your test infrastructure...but maybe that's for another conversation. 🙂)

I can elaborate a bit: We have hundreds (closer to 1000) dlls in a directory for our server product, and a bit less for our client. These are pretty dynamic and depend on the configuration for the particular customer. We also have a number of tool dlls in a separate directory.

Our harvesting system than generates nunit files for each directory. For the server and client we use a single domain model, which saves a lot of time and memory (and since those dlls would normally coexist anyway). For tool, we use a multiple process model so we can ensure that NUnit reads the .dll.config file for each .dll separately (as these would not normally coexist).

The regression causes the process model to be single for those, which disables .config file reading (as per design) and thus fails all the tests (as dll forwards are no longer respected).

The reason it's hard for us to use command line options is that we'd have to add that into our harvesting and test execution infrastructure, which currently only supports passing around nunit files and filters (--test).

Anyway, long rant, but TL;DR: we have a pretty elaborate test infrastructure. My work around will instead be a custom release of NUnit, which I have done several time before (while waiting for upstream patches to release).

**Comment by ChrisMaddock** [comment]
@cyanite That's really cool to hear. 😄 This is exactly what the console/engine is designed for and capable of, but we don't hear of many people who are using it to it's full potential in this way.

I'm glad all the additional complexity we have around this is justified for usecases like yours. (Now we just need to make it more stable. :-p) Thanks for sharing - will have a look at a release at the weekend. 🙂

**Comment by ChrisMaddock** [comment]
Thanks for picking this up @cyanite. v3.11.1 is now released on GitHub and NuGet. Chocolatey has temporarily locked my account (🤦‍♂) but I'll get round to that at some point in the next week.

Thanks for you help!

**Comment by CharliePoole** [comment]
@ChrisMaddock Do you want me to upload the chocolatey packages for you?

**Comment by ChrisMaddock** [comment]
That would be great Charlie - thanks! I uploaded them to the GitHub release already. 😄

**Comment by CharliePoole** [comment]
@ChrisMaddock I downloaded the nunit-console-runner package and attempted to upload it to chocolatey.org. I got their "Oh No we broke something" page, which is more cute than it is useful.

It doesn't say there's anything wrong with the package or with my credentials, however, so maybe I can just try again later in case something is down.

**Comment by ChrisMaddock** [comment]
Oh. 🙁 Thanks anyway Charlie. I’ll try again tomorrow too.

**Comment by CharliePoole** [comment]
Just tried again. No luck. :disappointed:

**Comment by ChrisMaddock** [comment]
Same error message today. I've tweeted them to make sure they're aware. 🙂

**Comment by CharliePoole** [comment]
I had the same message with a   TestCentric package. I asked about it on gitter.

**Comment by CharliePoole** [comment]
@ChrisMaddock The error I had is a known issue of the Gui and the solution is to use `choco push` from the command-line. I did it with my testcentric-gui package and it's now in review. I'll wait till that passes before trying yours I think.

If you want to do it your self, use choco apikey command at an administrator prompt to set up your key. You can do the pushing from a normal prompt.

**Comment by ChrisMaddock** [comment]
Thanks Charlie. I won’t be able to get back to a desktop till Tuesday at the earliest now - so would appreciate your help if you can push, otherwise I’ll sort it later in the week. 😊

**Comment by CharliePoole** [comment]
My TestCentric package went through so I pushed nunit-console to chocolatey as well. We should start seeing the verification messages shortly.

**Comment by CharliePoole** [comment]
It's published now, starting verification.

**Comment by ChrisMaddock** [comment]
Great - thanks Charlie!
